server/src/main/java/org/elasticsearch/index/mapper/MapperService.java

```diff
@@ -593,6 +593,11 @@ public DocumentMapperForType documentMapperWithAutoCreate(String type) {
      * Given the full name of a field, returns its {@link MappedFieldType}.
      */
     public MappedFieldType fieldType(String fullName) {
+        if (fullName.equals(TypeFieldMapper.NAME)) {
+            String type = mapper == null ? null : mapper.type();
+            return new TypeFieldMapper.TypeFieldType(type);
+        }
```

Comment:
for my understanding: does this mean that the field type does not come back with the below field type lookup? Did it before?

Reply:
It does come back with the lookup, but it may not have the correct type set on it (for example, if the index in question did not have any mappings set when the MapperService was originally built).  This way we know we always have the correct type.

Reply:
ok, so could there be places where the type field mapper is retrieved in some way that could expose the wrong type then?

Reply:
From a quick review:
- the only other way to access field types is via `MapperService#documentMapper#fieldTypes()` directly
- this is exposed via `MapperService#fieldTypes()`
- `MapperService#fieldTypes()` is only called by the index warming code, and the type field does not use global ordinals so is unaffected

So I think we're safe.  We should probably remove `MapperService#fieldTypes()` at some point though, as it is indeed asking for trouble to expose this to the outside world.

Reply:
ok thanks for double checking

Reply:
Thanks for keeping me honest :)